Adam Hyperparameter Tuning/TextImageUnfrozen_Adam16.ipynb

```diff
@@ -0,0 +1,4324 @@
+    "\n",
+    "losses_genre = []\n",
+    "losses_popularity = []\n",
+    "losses_danceability = []\n",
+    "losses_energy = []\n",
+    "\n",
+    "# Training loop\n",
+    "for epoch in range(num_epochs):\n",
+    "    for batch in range(num_training_batches):\n",
+    "        start = batch * batch_size\n",
+    "        end = min(start + batch_size, num_training_samples)\n",
+    "\n",
+    "        # Get data for this batch\n",
+    "        batch_X_sound = X_sound_train[start:end]\n",
+    "        # batch_X_album = X_album_train[start:end]\n",
+    "        batch_X_spec = X_spec_train[start:end]\n",
+    "        batch_X_spec = batch_X_spec.permute(0, 3, 1, 2)\n",
+    "\n",
+    "        batch_X_lyric = X_lyrics_train[start:end]\n",
+    "\n",
+    "        max_length = max([len(text_model.tokenizer.encode(text)) for text in batch_X_lyric])\n",
+    "        tokenized_sequences = [text_model.tokenizer.encode(text, padding='max_length', max_length=text_model.tokenizer.model_max_length, truncation=True) for text in batch_X_lyric]\n",
+    "\n",
+    "        # Generate attention masks\n",
+    "        attention_masks = [[1 if token != 0 else 0 for token in seq] for seq in tokenized_sequences]\n",
+    "\n",
+    "        # Convert tokenized sequences and attention masks to tensors and move to device\n",
+    "        batch_X_lyric = torch.tensor(tokenized_sequences, dtype=torch.long).to(device)\n",
+    "        attention_masks = torch.tensor(attention_masks, dtype=torch.long).to(device)\n",
+    "\n",
+    "        batch_y_genre = y_genre_train[start:end]\n",
+    "        batch_y_popularity = y_popularity_train[start:end]\n",
+    "        batch_y_danceability = y_danceability_train[start:end]\n",
+    "        batch_y_energy = y_energy_train[start:end]\n",
+    "\n",
+    "\n",
+    "        # Zero the gradients\n",
+    "        optimizer_sound.zero_grad()\n",
+    "        # optimizer_album.zero_grad()\n",
+    "        optimizer_spec.zero_grad()\n",
+    "\n",
+    "\n",
+    "        optimizer_genre.zero_grad()\n",
+    "        optimizer_popularity.zero_grad()\n",
+    "        optimizer_danceability.zero_grad()\n",
+    "        optimizer_energy.zero_grad()\n",
+    "\n",
+    "        # Retrieve the four outputs given our embeddings\n",
+    "        sound_embedding = sound_model(batch_X_sound)\n",
+    "        # album_embedding = album_model(batch_X_album)\n",
+    "        spec_embedding = spec_model(batch_X_spec)\n",
+    "        text_embedding = text_model(batch_X_lyric, attention_masks)\n",
+    "\n",
+    "\n",
+    "        genre_output, popularity_output, danceability_output, energy_output = overall_model(sound_embedding,\n",
+    "                                                                        spec_embedding,\n",
+    "                                                                        text_embedding)\n",
+    "\n",
+    "        genre_output = genre_output.to(device)\n",
```

Comment:
Shouldn't these already be on the device since the models themselves are?

Reply:
the models are also stored on the gpu im pretty sure.